### PR TITLE
Voigt component as expression

### PR DESCRIPTION
**Pull request by jlaehne**
### Description of the change
Implemented Voigt lineshape component as expression based on the implementation of @AEljarrat  mentioned in #2251. As the existing `Voigt` function had specific features for photoemission spectroscopy, this has been moved to `PESVoigt`, but similar to the solution for the `Polynomial` component, this is still available as legacy `Voigt` component (with a deprecation warning) and the new component is at the moment only chosen for `legacy=False`.

### Progress of the PR
- [X] Change implemented,
- [X] update docstring,
- [x] update user guide,
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
>>> import hyperspy.api as hs
>>> import numpy as np
>>> voigt = hs.model.components1D.Voigt(legacy=False)
>>> 
```
Note that this example can be useful to update the user guide.

### CODE REVIEW

File: hyperspy/_components/pes_voigt.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    """Voigt lineshape.
+
+    The voigt peak is the convolution of a Lorentz peak with a Gaussian peak.
+
+    The formula used to calculate this is::
+
+        z(x) = (x + 1j gamma) / (sqrt(2) sigma)
+        w(z) = exp(-z**2) erfc(-1j z) / (sqrt(2 pi) sigma)
+
+        V(x) = scale Re(w(z(x-center)))
```
Comment (ericpre) [inline hyperspy/_components/pes_voigt.py]: typo: `scale * Re(w(z(x-center)))`

File: hyperspy/_components/pes_voigt.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    return scale * V.real
+
+
+class Voigt(Component):
+    # Legacy class to be removed in v2.0
+
+    """This is the legacy Voigt profile component dedicated to photoemission 
+    spectroscopy data analysis that will renamed to `PESVoigt` v2.0. To use
+    the new Voigt lineshape component set `legacy=False`. 
+
```
Comment (ericpre) [inline hyperspy/_components/pes_voigt.py:93]: Add:
```
See the documentation of ``hyperspy._components.voigt.Voigt`` to use the new definition of the Voigt component. 
```

File: hyperspy/_components/voigt.py [hunk 3]
```diff
@@ -16,136 +16,99 @@
...
-    """Voigt lineshape.
+class Voigt(Expression):
 
-    The voigt peak is the convolution of a Lorentz peak with a Gaussian peak.
+    r"""Voigt component.
 
-    The formula used to calculate this is::
+    Symmetric peak shape based on the convolution of a Lorentzian and Normal 
+    (Gaussian) distribution. In this case using an approximate formula by David 
+    (see Notes).
```
Comment (ericpre) [inline hyperspy/_components/voigt.py]: Is it worth mentioning the difference with a pseudo-Voigt? Linear combination of Gaussian and Lorentzian versus this approximation for the convolution of Gaussian and Lorentzian?

File: hyperspy/_components/voigt.py [hunk 4]
```diff
@@ -171,58 +134,54 @@ def estimate_parameters(self, signal, E1, E2, only_current=False):
...
-        >>> s = hs.signals.Signal1D({'data' : data})
-        >>> s.axes_manager.axes[-1].offset = -10
-        >>> s.axes_manager.axes[-1].scale = 0.01
-        >>> g.estimate_parameters(s, -10,10, False)
+        >>> g = hs.model.components1D.Voigt(legacy=False)
+        >>> x = np.arange(-10, 10, 0.01)
+        >>> data = np.zeros((32, 32, 2000))
+        >>> data[:] = g.function(x).reshape((1, 1, 2000))
+        >>> s = hs.signals.Signal1D(data)
+        >>> s.axes_manager._axes[-1].offset = -10
```
Comment (ericpre) [inline hyperspy/_components/voigt.py]: Why using `_axes` instead of `axes`?

File: hyperspy/tests/component/test_pesvoigt.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+@pytest.mark.parametrize(("lazy"), (True, False))
+@pytest.mark.parametrize(("only_current", "binned"), TRUE_FALSE_2_TUPLE)
+def test_estimate_parameters_binned(only_current, binned, lazy):
+    s = Signal1D(np.empty((200,)))
+    s.metadata.Signal.binned = binned
+    axis = s.axes_manager.signal_axes[0]
+    axis.scale = .05
+    axis.offset = -5
+    g1 = PESVoigt()
```
Comment (ericpre) [inline hyperspy/tests/component/test_pesvoigt.py:53]: Add test for `Voigt(legacy=True)`?

File: hyperspy/_components/voigt.py [hunk 6]
```diff
@@ -16,136 +16,99 @@
...
-def voigt(x, FWHM=1, gamma=1, center=0, scale=1):
-    """Voigt lineshape.
+class Voigt(Expression):
 
-    The voigt peak is the convolution of a Lorentz peak with a Gaussian peak.
+    r"""Voigt component.
 
-    The formula used to calculate this is::
+    Symmetric peak shape based on the convolution of a Lorentzian and Normal 
+    (Gaussian) distribution. In this case using an approximate formula by David
```
Comment (ericpre) [inline hyperspy/_components/voigt.py]: Because the formula to calculate the Voigt component looks complicated at first read, I would first give the definition of the Voigt function with the convolution (as a reference) and then I would give the formula used for the calculation.

File: hyperspy/_components/voigt.py [hunk 7]
```diff
@@ -16,136 +16,99 @@
...
-    Ref: W.I.F. David, J. Appl. Cryst. (1986). 19, 63-64
 
-    adjusted to use stddev and HWHM rather than FWHM parameters
+    Parameters
+    -----------
+    centre : float
+        Location of the maximum of the peak.
+    area : float
+        Intensity below the peak.
+    lwidth : float
```
Comment (ericpre) [inline hyperspy/_components/voigt.py]: Any reason to use `lwidth` and `gwidth` instead of `gamma` and `sigma`?

**Review by ericpre — COMMENTED** [review]
Nice PR, thanks @jlaehne. I made a couple of suggestion to clarify the docstring.
It would be good to add the new Voigt component to the `remove_background` tool. Other than that, this PR looks good to me.

@francisco-dlp, could you have a quick look at the API change and the legacy component, it looks fine to me, but a second pair of eyes would be too good to make sure this is done correctly.

File: hyperspy/_components/pes_voigt.py (hunk 1 above)
Comment (jlaehne) [reply hyperspy/_components/pes_voigt.py]: I changed it to math typesetting like in the new Voigt.

File: hyperspy/_components/pes_voigt.py (hunk 2 above)
Comment (jlaehne) [reply hyperspy/_components/pes_voigt.py:93]: done

File: hyperspy/_components/voigt.py (hunk 3 above)
Comment (jlaehne) [reply hyperspy/_components/voigt.py]: I added an explanation based on the abstract of David et al.

File: hyperspy/tests/component/test_pesvoigt.py (hunk 5 above)
Comment (jlaehne) [reply hyperspy/tests/component/test_pesvoigt.py:53]: added test

File: hyperspy/_components/voigt.py (hunk 4 above)
Comment (jlaehne) [reply hyperspy/_components/voigt.py]: It was copy pasted from `Gaussian`. I also removed it in `Gaussian`, `GaussianHF`, and `Lorentzian`.

File: hyperspy/_components/voigt.py (hunk 6 above)
Comment (jlaehne) [reply hyperspy/_components/voigt.py]: good idea. done.

**Comment by jlaehne** [comment]
Concerning the use of the `lwidth` and `gwidth` parameters instead of `gamma` and `sigma`:
I took it from the component of @AEljarrat. 

However, I also find it more intuitive to work with a width parameter than with `sigma`. Of course, `gamma=HWHM=lwidth`, but then I thought it more consistent to use a similar keyword for both distributions. 

Would you prefer `gamma` and `FWHM` like in the PESVoigt?

Maybe even better would be `gamma` and `fwhm` to be consistent with `Lorentzian` and `GaussianHF` components?

@ericpre, @francisco-dlp - what is your opinion on the naming?

(PS: Also added Voigt to `remove_background`)

File: hyperspy/_components/voigt.py (hunk 7 above)
Comment (jlaehne) [reply hyperspy/_components/voigt.py]: see comment  below

File: hyperspy/_components/voigt.py (hunk 4 above)
Comment (ericpre) [reply hyperspy/_components/voigt.py]: Sorry for this wrong suggestion, there is no `axes` attribute to the axes manager! I made a PR to this branch to fix it: https://github.com/jlaehne/hyperspy/pull/2.
I guess, this is one of these where one writes a bit too quickly and the brain is following up properly... ;)

**Comment by ericpre** [comment]
Regarding `fwhm` versus `sigma`, I don't have any particular view and indeed trying to make consistent with other function would be sensible!
- `Gaussian` has `A`, `sigma` and `centre`
- `GaussianHF` has `height`, `FWHM` and `centre`
- `Lorentzian` has `A`, `gamma` and `centre`
- `PSEVoigt` has `area`, `FWHM`, `gamma` and `centre`

To be consistent with `Gaussian` and `Lorentzian` I would say that `sigma` and `gamma` should be used in `Voigt`.

Other opinions?

File: hyperspy/signal_tools.py [hunk 8]
```diff
@@ -1003,12 +1005,15 @@ def set_background_estimator(self):
...
-            self.bg_line_range = 'full'
+        elif self.background_type == 'Power Law':
+            self.background_estimator = components1d.PowerLaw()
+            self.bg_line_range = 'from_left_range'
         elif self.background_type == 'SkewNormal':
             self.background_estimator = components1d.SkewNormal()
             self.bg_line_range = 'full'
+        elif self.background_type == 'Voigt':
+            with ignore_warning(message="The API of the `Voigt` component"):
+                self.background_estimator = components1d.Voigt()
```
Comment (ericpre) [inline hyperspy/signal_tools.py]:
```suggestion
                self.background_estimator = components1d.Voigt(legacy=False)
```
to be consistent with the `remove_background` method.

Comment (jlaehne) [reply hyperspy/signal_tools.py]: Thanks. I changed that.

File: hyperspy/_components/pes_voigt.py [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+    centre: Parameter
+    FWHM : Parameter
+    gamma : Parameter
+    resolution : Parameter
+    shirley_background : Parameter
+    non_isochromaticity : Parameter
+    transmission_function : Parameter
+    spin_orbit_splitting : Bool
+    spin_orbit_branching_ratio : float
+    spin_orbit_splitting_energy : float
```
Comment (dnjohnstone) [inline hyperspy/_components/pes_voigt.py]: I realise that this is due to become legacy, but shouldn't the attributes that are specific to PESVoigt not be in this more general Voigt docstring?

File: hyperspy/_components/pes_voigt.py [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+        FWHM=:math:`2 \sigma \sqrt(2 \log(2))` of the Gaussian distribution.
+    gamma : Parameter
+        :math:`\gamma` of the Lorentzian distribution.
+    resolution : Parameter
+    shirley_background : Parameter
+    non_isochromaticity : Parameter
+    transmission_function : Parameter
+    spin_orbit_splitting : Bool
+    spin_orbit_branching_ratio : float
+    spin_orbit_splitting_energy : float
```
Comment (dnjohnstone) [inline hyperspy/_components/pes_voigt.py:191]: I think these parameters need some explanation.

File: hyperspy/_components/pes_voigt.py [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+from hyperspy._components.gaussian import _estimate_gaussian_parameters
+
+
+sqrt2pi = math.sqrt(2 * math.pi)
+sigma2fwhm = 2 * math.sqrt(2 * math.log(2))
+
+
+def voigt(x, FWHM=1, gamma=1, center=0, scale=1):
+    r"""Voigt lineshape.
+
```
Comment (dnjohnstone) [inline hyperspy/_components/pes_voigt.py:32]: stray 'r' here

File: hyperspy/_components/pes_voigt.py [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+        self.transmission_function.free = False
+        # Options
+        self.shirley_background.active = False
+        self.spin_orbit_splitting = False
+        self.spin_orbit_branching_ratio = 0.5
+        self.spin_orbit_splitting_energy = 0.61
+        self.isbackground = False
+        self.convolved = True
+
+    def function(self, x):
```
Comment (dnjohnstone) [inline hyperspy/_components/pes_voigt.py:229]: Doesn't this need a docstring?

File: hyperspy/_components/pes_voigt.py [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+        x2 : float
+            Defines the right limit of the spectral range to use for the
+            estimation.
+
+        only_current : bool
+            If False estimates the parameters for the full dataset.
+
+        Returns
+        -------
+        bool
```
Comment (dnjohnstone) [inline hyperspy/_components/pes_voigt.py]: This seems like a very strange return for a method that is called estimate parameters - I understand why it is, but I think it should explain here.

File: hyperspy/_components/pes_voigt.py [hunk 14]
```diff
@@ -0,0 +1,331 @@
+# -*- coding: utf-8 -*-
+# Copyright 2007-2016 The HyperSpy developers
```
Comment (dnjohnstone) [inline hyperspy/_components/pes_voigt.py]: probably needs an update more generally

File: hyperspy/_components/voigt.py [hunk 15]
```diff
@@ -16,136 +16,110 @@
...
 sqrt2pi = math.sqrt(2 * math.pi)
+sigma2fwhm = 2 * math.sqrt(2 * math.log(2))
 
 
-def voigt(x, FWHM=1, gamma=1, center=0, scale=1):
-    """Voigt lineshape.
+class Voigt(Expression):
 
-    The voigt peak is the convolution of a Lorentz peak with a Gaussian peak.
+    r"""Voigt component.
```
Comment (dnjohnstone) [inline hyperspy/_components/voigt.py:34]: stray r

**Review by dnjohnstone — COMMENTED** [review]
This basically looks good to me except for some fairly superficial housekeeping and the conflict on updates to remove_background

File: hyperspy/_components/voigt.py [hunk 16]
```diff
@@ -16,136 +16,110 @@
...
 
-    The formula used to calculate this is::
+    Symmetric peak shape based on the convolution of a Lorentzian and Normal 
+    (Gaussian) distribution:
+    
+        .. math::
```
Comment (ericpre) [inline hyperspy/_components/voigt.py]: Indentation is not correct here and it makes the following formula not to render properly.

File: hyperspy/_components/pes_voigt.py (hunk 11 above)
Comment (ericpre) [reply hyperspy/_components/pes_voigt.py:32]: This `r` is necessary to render the latex properly.

File: hyperspy/_components/pes_voigt.py (hunk 10 above)
Comment (ericpre) [reply hyperspy/_components/pes_voigt.py:191]: Indeed, but since this is simply moving code around, this can be optional for PR.

File: hyperspy/_components/pes_voigt.py (hunk 12 above)
Comment (ericpre) [reply hyperspy/_components/pes_voigt.py:229]: Same as above, since this is simply moving code around, this can be optional for PR.

File: hyperspy/_components/pes_voigt.py (hunk 13 above)
Comment (ericpre) [reply hyperspy/_components/pes_voigt.py]: A quick search in the code shows that it is actually used: https://github.com/hyperspy/hyperspy/blob/ff2d9d38098831d0e2cde1239dd80bff056b9f67/hyperspy/signal_tools.py#L1089-L1095

As usual, documenting it would be good indeed.

File: hyperspy/_components/voigt.py (hunk 15 above)
Comment (ericpre) [reply hyperspy/_components/voigt.py:34]: Same as above, this `r` is necessary.

File: hyperspy/_components/voigt.py [hunk 17]
```diff
@@ -171,58 +145,62 @@ def estimate_parameters(self, signal, E1, E2, only_current=False):
...
+        >>> data[:] = g.function(x).reshape((1, 1, 2000))
+        >>> s = hs.signals.Signal1D(data)
+        >>> s.axes_manager[-1].offset = -10
+        >>> s.axes_manager[-1].scale = 0.01
+        >>> g.estimate_parameters(s, -10, 10, False)
 
         """
         super(Voigt, self)._estimate_parameters(signal)
         axis = signal.axes_manager.signal_axes[0]
+        centre, height, sigma = _estimate_gaussian_parameters(signal, x1, x2,
```
Comment (ericpre) [inline hyperspy/_components/voigt.py:161]: Only the `sigma` is estimated here which make it fairly inaccurate expect when the degree of gaussianity is high (in which case, it may be better to use a gaussian!).
If someone could find a more accurate way to estimate the `gamma` value, that would be great, I started to look at this in the literature, but I haven't successful so far.
Even if this is quite critical, we can leave to another PR.

File: hyperspy/signal_tools.py [hunk 18]
```diff
@@ -1003,12 +1004,15 @@ def set_background_estimator(self):
...
-            self.bg_line_range = 'full'
+        elif self.background_type == 'Power Law':
+            self.background_estimator = components1d.PowerLaw()
+            self.bg_line_range = 'from_left_range'
         elif self.background_type == 'SkewNormal':
             self.background_estimator = components1d.SkewNormal()
             self.bg_line_range = 'full'
+        elif self.background_type == 'Voigt':
+            with ignore_warning(message="The API of the `Voigt` component"):
+                self.background_estimator = components1d.Voigt(legacy=False)
```
Comment (ericpre) [inline hyperspy/signal_tools.py:1015]:
```suggestion
                self.background_estimator = components1d.Voigt(legacy=False)
           self.bg_line_range = 'full'
```

File: hyperspy/_components/pes_voigt.py (hunk 9 above)
Comment (jlaehne) [reply hyperspy/_components/pes_voigt.py]: OK, I'll replace it by a note that refers to the two respective components. Anyway, it is called `Parameters` in other components ;-)

File: hyperspy/_components/pes_voigt.py (hunk 10 above)
Comment (jlaehne) [reply hyperspy/_components/pes_voigt.py:191]: Yes, but as I have not programmed the PES component, I'll leave it to someone more into the topic to add further explanations.

File: hyperspy/_components/pes_voigt.py (hunk 13 above)
Comment (jlaehne) [reply hyperspy/_components/pes_voigt.py]: I'll add a note, if we agree on it, I can also add it to other components that are used by `remove_background`.

File: hyperspy/_components/pes_voigt.py (hunk 14 above)
Comment (jlaehne) [reply hyperspy/_components/pes_voigt.py]: Yes, I can change it in the files used here, but it can be found across the project.

File: hyperspy/_components/voigt.py (hunk 16 above)
Comment (jlaehne) [reply hyperspy/_components/voigt.py]: good catch

File: hyperspy/_components/voigt.py (hunk 17 above)
Comment (jlaehne) [reply hyperspy/_components/voigt.py:161]: I agree, but I also have not found a good solution so far, thus I propose to merge this PR and leave it. You could open an issue if you deem it appropriate.

File: hyperspy/signal_tools.py (hunk 18 above)
Comment (jlaehne) [reply hyperspy/signal_tools.py:1015]: done, thanks

File: hyperspy/_components/pes_voigt.py (hunk 11 above)
Comment (dnjohnstone) [reply hyperspy/_components/pes_voigt.py:32]: Ah - my bad, reviewing = learning = :)
